src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+    def __repr__(self):
+        location = [p.name for p in self.path]
+        location = '.'.join(location)
+        return f"<{self.__class__} paths={location}>"
+
+
+def dal_resolver(ctx, deferred_dal_attr_request):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    service_or_method = None
+    for (_path, _service_or_method) in resolve(ctx, deferred_dal_attr_request.path):
+        if not _service_or_method:
+            raise DalMethodError(path=tuple(_path))
+        elif not isinstance(_service_or_method, (Callable, Service)):
+            raise DalMethodError(path=tuple(_path))
+        service_or_method = _service_or_method
+    return service_or_method
+
+
+def resolve(ctx: DataAccessContext, path: Tuple[PathSegment]):
+    """
+    A generator that returns a list of path segments traversed
+    during resolving the dal attribute and the method to call, if any.
+    """
+    service_or_method = ctx.dal._services  # noqa
+    paths = list(path)
+    location = []
+
+    while paths:
+        path_segment = paths.pop(0)
+        location.append(path_segment)
+        # Third time through, service_or_method might be `None`, but we want
+        # to continue walking the path. Everything after the first missing
+        # service/method will be `(location, None)`.
+        if service_or_method:
+
+            # This if condition is handling the case of the first loop here.
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                except AttributeError:
+                    service_or_method = None
+
+            yield location, service_or_method
+        else:
+            yield location, None
+
+
+def dal_method_resolver_middleware(request: DalMethodRequest, handler: Callable):
+    """
+    A Method Middleware that resolves deferred dal attribute access to
+    underlying services/methods that has to be called
+    Args:
+         request: Input from caller
+         handler: Downstream middleware or actual DAL method handler
+    """
+    request.dal_method = dal_resolver(request.ctx, DeferredDalAttrRequest(handler, request.path))
```

Comment:
```suggestion
    request.dal_method = dal_resolver(request.ctx, request.path)
```

I think this requires some changes to `dal_resolver` but:

1. dal_resolver doesn't need the handler. It's job is to resolve a path.
2. We end up creating and tearing down and recreate a DeferredDalAttrRequest multiple times if done this way.